// @tldraw/sync-core/dist-esm/index.mjs
import { registerTldrawLibraryVersion } from "@tldraw/utils";
import { ClientWebSocketAdapter, ReconnectManager } from "./lib/ClientWebSocketAdapter.mjs";
import { RoomSessionState } from "./lib/RoomSession.mjs";
import { TLRemoteSyncError } from "./lib/TLRemoteSyncError.mjs";
import { TLSocketRoom } from "./lib/TLSocketRoom.mjs";
import {
  TLCloseEventCode,
  TLSyncClient
} from "./lib/TLSyncClient.mjs";
import {
  DocumentState,
  TLSyncRoom
} from "./lib/TLSyncRoom.mjs";
import { chunk } from "./lib/chunk.mjs";
import {
  RecordOpType,
  ValueOpType,
  applyObjectDiff,
  diffRecord,
  getNetworkDiff
} from "./lib/diff.mjs";
import {
  TLIncompatibilityReason,
  getTlsyncProtocolVersion
} from "./lib/protocol.mjs";
registerTldrawLibraryVersion(
  "@tldraw/sync-core",
  "3.2.0",
  "esm"
);
export {
  ClientWebSocketAdapter,
  DocumentState,
  ReconnectManager,
  RecordOpType,
  RoomSessionState,
  TLCloseEventCode,
  TLIncompatibilityReason,
  TLRemoteSyncError,
  TLSocketRoom,
  TLSyncClient,
  TLSyncRoom,
  ValueOpType,
  applyObjectDiff,
  chunk,
  diffRecord,
  getNetworkDiff,
  getTlsyncProtocolVersion
};
